### DISCUSSION TITLE
Hot to call a method with argument array as input

### DISCUSSION THREAD

**Discussion by uioplmn** [FreeOpcUa/python-opcua#1320] [Q&A]
Hi. I am new to OPC UA and stuck with calling a method with an argument array as input. I posted the details of my question on [Stack Overflow](https://stackoverflow.com/questions/67604952/python-opc-ua-call-method-with-array-input-argument).

Can anybody give me a hint? Thanks!

**Comment by AndreasHeine** [comment]
you should not pass a list of values instead you should pass a list of ua.Variant classes with the right types of each like shown in the method input arguments (e.g. UaExpert)

**Reply by uioplmn** [reply]
Ok thanks, that helps a lot already! Can you give me some more instructions how I can create the classes, given the input arguments are characterized like this:

![](https://i.stack.imgur.com/hyh7s.png)

![](https://i.stack.imgur.com/VY08s.png)

**Reply by AndreasHeine** [reply]
```python
import opcua
from opcua import Client, ua

client = Client("opc.tcp://127.0.0.1:4840/UA")
client.set_user = "user"
client.set_password = "pw"  

client.connect()

parent = client.get_node("ns=4;i=1001")
method = client.get_node("ns=4;i=1002")
result = parent.call_method(method,
    ua.Variant(5, ua.VariantType.UInt32),
    ua.Variant(99, ua.VariantType.UInt32)
)

print(result)

client.disconnect()
```

> 
> C:\Users\andre>python c:/Users/andre/client.py
> Requested secure channel timeout to be 3600000ms, got 600000ms instead
> Requested session timeout to be 600000ms, got 3000000ms instead
> [5, 99, 100, 200, 300]
> 
> C:\Users\andre>

![Unbenannt](https://user-images.githubusercontent.com/56362817/119045245-794c5280-b9bb-11eb-9628-cbf7f0e49b8e.PNG)

**Reply by uioplmn** [reply]
That did the trick. Thanks a lot for your help!

**Reply by h0tnanny** [reply]
Hey! I had a similar problem, but the solution presented above did not help
![screenshot](https://user-images.githubusercontent.com/47948448/157306909-47145f4a-9476-497a-bfc7-25253e24ca4c.png)
![input arguments method](https://user-images.githubusercontent.com/47948448/157307577-43c72b2e-4a07-43e8-bd63-b96ca5e2ef4c.png)

**Reply by AndreasHeine** [reply]
because your inputargument is an extensionobject (ScanSettings)!

**Reply by h0tnanny** [reply]
Thanks, can you please tell me how to correctly pass the input parameters?
![arg](https://user-images.githubusercontent.com/47948448/157309740-f28c62c5-7cca-4474-8117-7abb81a85bcd.png)

**Reply by AndreasHeine** [reply]
your client need to load the typedefinitions and then create a instance of that definition/class and pass it as value! but this library is deprecated not sure if it works you should switch to asyncua.

**Reply by h0tnanny** [reply]
Thank you very much, I will try!

**Comment by KolbenKeule** [comment]
Hi, I have a big problem with VariantType.ExtensionObject and install asyncua but I don´t know how it´s work. Sry I´m new with OPC-UA too :( I search for some weeks for a solution but nothing works.
![Code](https://user-images.githubusercontent.com/86122745/158404244-3cc84816-43e7-4d70-80f0-a7b424aab377.jpg)
![UAExpert](https://user-images.githubusercontent.com/86122745/158404309-7361235f-faa1-4ce1-a383-4270f27b09ba.jpg)

**Reply by AndreasHeine** [reply]
@KolbenKeule 

[](https://github.com/FreeOpcUa/python-opcua#please-switch-to--opcua-asyncio-which-also-has-a-sync-wrapper-with-very-few-changes-in-api)The library is deprecated.
Please switch to : [opcua-asyncio](https://github.com/FreeOpcUa/opcua-asyncio) which also has a [sync-wrapper](https://github.com/FreeOpcUa/opcua-asyncio/blob/master/asyncua/sync.py), with very few changes in API**

**Reply by KolbenKeule** [reply]
Hi, thank you for the fast answer.
I got opcua-asyncio running and found a good example for calling a method. But I get struggel with the nodes. I have no idea how can fix the problems and run the code.
![asycio part1](https://user-images.githubusercontent.com/86122745/158564253-1aa3fbdb-19d3-4bee-9237-55d3225698f2.jpg)
![asynio part2](https://user-images.githubusercontent.com/86122745/158564270-8b49f8d1-cded-4752-a699-e78e70e19355.jpg)
Can you help me again please?

**Reply by AndreasHeine** [reply]
i pinged you yesterday from https://github.com/FreeOpcUa/opcua-asyncio/discussions/839 on how to use ExtensionObjects!

**Comment by KolbenKeule** [comment]
Thank you very much. I only have a problem to understand this completely :(

**Reply by AndreasHeine** [reply]
https://reference.opcfoundation.org/v104/Core/docs/Part6/5.1.5/

**Comment by SPECTERBLUE** [comment]
Could someone help me out with this?

[https://stackoverflow.com/questions/72367498/how-to-call-a-method-with-multipuls-arguments-with-different-datatypes-as-input](url)

**Comment by SneOG1** [comment]
Hi, 
I am also a total beginner with OPC UA. I have a kinda similar problem and tried the suggested solution, but the error consists.

I am trying to use call method with two Integer arguments and i am receiving a "BadInvalidArgument"-error.
The code for the method call is as follows:

```
# Calling a method
        parent = await root.get_child(["0:Objects", "3:ServerInterfaces", "4:Server interface_1", "4:Methods-OPCUA_DB"])
        method = client.get_node("ns=4; i=5")

        input_arg1 = ua.Variant(5, ua.VariantType.UInt32)
        input_arg2 = ua.Variant(4, ua.VariantType.UInt32)

        res = await parent.call_method(method, input_arg1, input_arg2)
    
        print(f"Calling ServerMethod returned {res}")

```
But i receive following error:
```
Exception has occurred: BadInvalidArgument
One or more arguments are invalid.(BadInvalidArgument)
  File "H:\Development\Test_OPC_UA\OPC_UA_Client_test_async.py", line 43, in main
    res = await parent.call_method(method, input_arg1, input_arg2)
  File "H:\Development\Test_OPC_UA\OPC_UA_Client_test_async.py", line 49, in <module>
    asyncio.run(main())
asyncua.ua.uaerrors._auto.BadInvalidArgument: One or more arguments are invalid.(BadInvalidArgument)
```
I am able to call the method in UAExpert. The Server structure is as follows:
<img width="266" alt="image" src="https://github.com/FreeOpcUa/opcua-asyncio/assets/167765683/a1ad7f65-2829-43a1-b2c3-10f2826870b1">

and the attributes of the InputParameters are like this:
<img width="290" alt="image" src="https://github.com/FreeOpcUa/opcua-asyncio/assets/167765683/bca23a7b-0627-462c-bf9f-c6fc6cb76801">

**Comment by prlksk** [comment]


**Reply by schroeder-** [reply]
Error code?

**Reply by prlksk** [reply]
When I have this:     result = device.call_method(method, ua.Variant(0, ua.VariantType.UInt32))
Error: 
opcua.ua.uaerrors._auto.BadDecodingError: "Decoding halted because of invalid data in the stream."(BadDecodingError)

when I import : from enum import IntEnum
from opcua import ua
opcua.ua.uaerrors._auto.BadInvalidArgument: "One or more arguments are invalid."(BadInvalidArgument)

**Reply by prlksk** [reply]
![Screenshot 2024-06-05 135714](https://github.com/FreeOpcUa/python-opcua/assets/24990351/a498ab6c-60ea-477e-99d6-9b0536653160)
